### DISCUSSION TITLE
Proposal to drop custom volume format and use `.tar.gz`

### DISCUSSION THREAD

**Discussion by nagy** [seaweedfs/seaweedfs#2116] [Ideas]
I want to suggest a potentially controversial proposal to drop the custom volume
format and adopt a specifically crafted, but valid `.tar.gz` format.

The main motivation behind this is, that users have access to their data without
a custom made exporter tool because they can rely on tools that they already
know. The idea is that users can extract their data with a simple `tar xf 1.tar.gz`
 on that volume file. This way, if seaweedfs will not be around
eventually, people can still extract their data with any tool that supports the
extraction of `.tar.gz` files. I think this can give people more confidence in
adopting seaweedfs, since they dont need to study a custom format on how to
recover their data from it in case of an emergency.

Of course, random access to files and a small overhead per file is still crucial
and this is the reason for the self-crafted encoder.

The DEFLATE algorithm, which gzip uses, supports blocks of stored(uncompressed)
and compressed blocks, where the latter ships with its own huffman-table. Also
gzip supports multi-file streams, which indicate the start of a new stream with
their respective blocks and huffman-tables. Tar files as well as gzip files are
able to be concatenated and read back as if they were one file. This allows us
to append files to this volume file.

The idea is to store the same data that we store in our current header inside
the tar header format in their respective places and use compressed DEFLATE
blocks to keep the overhead low and use stored blocks for the actual payload. We
cut off the file via a multi-file separator. The crossing point of such a
multi-file stream can be recorded in the `.idx` files as starting points and
allow for random access to each of the files since no huffman-table data is
needed from before that point.

As you can see in the image below, most of the data in the header are
zero-bytes, therefore they are compressed well. In my manual testing I was able
to get to an overhead of around 53 Bytes per file, but perhaps this can even be
reduced some more. Keep in mind that, at the end of a tar "record" it is padded
to borders of 512 byte, which adds a bit more of overhead, but since those are
allowed to be zero bytes as well, they can be compressed well also. And finally,
due to the DEFLATE algorithm there are another 3 Bit of overhead for every 64KiB
of data.

If you find this interesting, I can write some proof of concept scripts to show
this on some files.

![tarheader](https://user-images.githubusercontent.com/692274/121219556-1d6f3e00-c884-11eb-935e-c9d034a8c564.png)

**Comment by chrislusf** [comment]
This seems a bold idea. But definitely needs some proof of concept built first.

Not sure about the gain though. Currently you can use `weed export` to convert the .dat file into a tar file already.

**Reply by nagy** [reply]
> But definitely needs some proof of concept built first.

I am happy to provide that.

> Currently you can use `weed export` to convert the .dat file into a tar file already.

I know, but for that you need to have a system that has a working installation
of seaweedfs. That might not always be the case.

Also you need to know that this `.dat` file can be exported via a `weed export`
command. The person creating the cluster and the one rescuing the cluster might
not be the same person. There may be years or decades apart. Sure, you can write
a note in the header of the file pointing to and describing the name of the
project, if this is not already the case.

But I would argue, that seeing the file ending `.tar.gz` is pretty explanatory
on the format where as a `.dat` file, not that much. Also it frees us from
having to produce and keep around a documentation for our custom format.

> Not sure about the gain though.

If you do not see value in any of the above then there is no need to go on with
this. Certainly, there is no performance gain here. While I would say that the
actual performance should be about the same, there are a dozen more bytes per
file overhead more than right now, but maybe only a bit less than a dozen.

**Reply by chrislusf** [reply]
This may be good to attract casual users. But SeaweedFS will lose the interest of enterprise users where each byte or millisecond matters,or lose to many closed source solutions.

**Comment by nagy** [comment]
Here is the POC in form of a shellscript.

```sh
# lets setup our environment
mkdir /tmp/empty
cd -- /tmp/empty
echo -ne T > e
echo -ne U > f

# lets see an empty gzip header
true | gzip | hexdump -vC

# Result:
# 00000000  1f 8b 08 00 00 00 00 00  02 03 03 00 00 00 00 00  |................|
# 00000010  00 00 00 00                                       |....|
# 00000014

# Create the first tar file. Tar writes an EOF block and the end of each file.
# This signals to tar that the file has ended and is only filled with zero
# bytes. We do not need this block and therefore letting only the first 1024
# bytes pass. This allows us to concatenate two tar files together.
tar --numeric-owner --owner=root:0 --group=root:0 --mtime="UTC 1970-01-01" -cf - e | dd bs=1 count=1024 | gzip > file1.tar.gz
hexdump -vC file1.tar.gz


# Result:
# 00000000  1f 8b 08 00 00 00 00 00  02 03 4b 65 a0 3d 30 00  |..........Ke.=0.|
# 00000010  02 33 13 13 30 0d 04 e8  34 10 18 22 b1 41 e2 66  |.3..0...4..".A.f|
# 00000020  06 26 c6 0c 0a 06 74 70  1b 43 69 71 49 62 91 82  |.&....tp.CiqIb..|
# 00000030  02 c3 08 05 21 0c a3 60  24 03 00 a6 12 7d fa 00  |....!..`$....}..|
# 00000040  04 00 00                                          |...|
# 00000043

# Create the second tar file
tar --numeric-owner --owner=root:0 --group=root:0 --mtime="UTC 1970-01-01" -cf - f | dd bs=1 count=1024 | gzip > file2.tar.gz
hexdump -vC file2.tar.gz

# Result:
# 00000000  1f 8b 08 00 00 00 00 00  02 03 4b 63 a0 3d 30 00  |..........Kc.=0.|
# 00000010  02 33 13 13 30 0d 04 e8  34 10 18 22 b1 41 e2 66  |.3..0...4..".A.f|
# 00000020  06 40 e5 0a 06 74 70 1b  43 69 71 49 62 91 82 02  |.@...tp.CiqIb...|
# 00000030  c3 08 05 a1 0c a3 60 24  03 00 c1 b6 77 e8 00 04  |......`$....w...|
# 00000040  00 00                                             |..|
# 00000042

# lets combine them and it is still readable
cat file{1,2}.tar.gz > both.tar.gz
rm -f -- e f
# and lets reextract them
tar -tvf both.tar.gz
tar -xf both.tar.gz
cat e
cat f

# Result:
# -rw-r--r-- 0/0               1 1970-01-01 01:00 e
# -rw-r--r-- 0/0               1 1970-01-01 01:00 f
# TU


# so the .idx files would save the position in between the files and therefore
# would be able to have direct access to the initial tar archive and therefore
# the inner file. Without much optimization the current overhead for such a
# one-byte file file is 65 bytes ( 0x42 - 1 byte file size ). You could add a
# "--mode=000" to the tar command to save some more bytes but that would make
# the unpacking phase somewhat uncomfortable, since all extracted files are not
# readable without a permission change. Also you can zero out any fields in the
# tar file that are not used, for example the user and group field. I did not do
# that here because for that you need to recalculate the tar header checksum.
#
# There is however another way how to save some more overhead. As it stands now,
# there is a new gzip stream for every file and this stream contains a DEFLATE
# body. This DEFLATE body contains a huffman table that compresses the rest of
# the data. We can combine multiple files into one gzip stream and therefore
# share the huffman data between them, leading to higher compression rates. Lets
# say we want to pack 4 files together. Maybe we have a scenario where there
# four files with 4MB each in size, resulting in a roughly 16MB gzip stream.
#
# Normally this gzip stream would need to be read from the start, disallowing us
# from having direct random access to files. But if my understanding serves me
# correctly, than there is another way of having random access to any file
# within that pack. And that is, that we also store the position of the huffman
# table of the DEFALTE body in the index. Then we can immediately decode any
# file within the whole stream. We might have to do two seeks for this though.
#
# This latter concept would deserve its own POC but that is just to reduce the
# number of overhead bytes anyway. So I think there is still some room for
# overhead shaving.

```
